docs/ADR-3-Dependencies-version-constraints-in-cabal-file.md

```diff
@@ -0,0 +1,87 @@
+# Status
+
+📜 Proposed 2023-05-15
+
+# Context
+
+In a project with multiple packages, in order to provide stable and reproducible builds, there is a need to
+control version bounds of the dependencies used.
+This is done by entering version bounds into the `.cabal` file, like described in [Cabal documentation](https://cabal.readthedocs.io/en/latest/cabal-package.html#pkg-field-build-depends).
+Let's consider an example where multiple package *components* (e.g. libraries, internal libraries, test-suites,
+executables) form a more complex dependency graphs like the following one:
+```mermaid
+graph LR
+ B --> A
+ C --> A
+ D --> B
+ D --> C
+```
+Considering an example case, where `B` and `C` but not `A`, have the same dependencies, we need to encode version ranges twice.
+This results in duplicated entries in `.cabal` file, with duplicated version bounds.
+
+# Considered options
+
+1. **Using explicit common stanzas, one per dependency.** See for example: https://github.com/input-output-hk/hedgehog-extras/blob/e7f3c9ff967ed6f3b4c8c17013a4e0c2f541e053/hedgehog-extras.cabal#L19
+    * **Advantages**
+      * Fine grained control over what dependency is included where, with the single place which defines
+        version bounds.
+    * **Disadvantages**
+      * A lot of boilerplate when defining new dependencies.
+
+1. **Moving common dependencies to `common` stanzas, multiple dependencies in one stanza**.
+  This allows for grouping of dependencies between components in order specify version bound only once.
+  For example the same `common` stanza could be reused in `B` and `C` in order to define version bounds in a
+  single place.
+    * **Advantages**
+      * Grouping multiple dependencies together introduces less boilerplate in comparison to the
+        *one-stanza-per-one-dependency* option.
+    * **Disadvantages**
+      * It may be hard to define logical grouping for few unrelated dependencies.
+      * In the worst case scenario we'll end up with *one-stanza-per-one-dependency* definition of
+        dependencies, when those will be used in different components.
```

Comment:
Disadvantage: Each common stanza will need to be named.  There is a risk the naming will be inconsistent between projects and confusion could occur around that inconsistency and bike-shedding may result from having to deal with those inconsistencies.